@grafana/ui: adds a virtualized options for the Select component

gillesdemey opened the pull request:
**What this PR does / why we need it**:

This PR introduces the option to use a virtualized list for the `<Select />` component. I've opted to make it optional instead of the default because there are a couple of things that aren't supported with the virtual list, most notably:

* The `maxHeight` property isn't supported properly and effectively sets the height of the drop-down menu to a fixed height.

* The `description` property for a `SelectableValue` isn't supported since those are variable height depending on how wide the `<Select />` component is.

Using the default `<Select />` component with anything near a few thousand options is currently _very_ slow and un-usable.

**Which issue(s) this PR fixes**:

We have some customers and users with a _very_ large number of Alerting groups and we use the default `<Select />` component to allow picking any one of them.

**Special notes for your reviewer**:

More than happy to take another approach to this or help figure out how to solve the aforementioned unsupported features so we can make this the default for all instances of the `<Select />` component.

gillesdemey committed:
adds a virtualized options for the Select component

obetomuniz commented:
Hey @gillesdemey, great work here! We have an escalation https://github.com/grafana/support-escalations/issues/4008 that will be directly affected by these changes when merged. 

In any case, while testing it against my problem, I noticed an issue related to the styling that you can see right below: 

Without `virtualized` property in the `<Select>` component:
<img width="579" alt="Screen Shot 2022-09-23 at 11 55 56" src="https://user-images.githubusercontent.com/1680157/191991400-736ad572-dfa6-4a91-8d84-902322457764.png">

With `virtualized` property in the `<Select>` component:
<img width="187" alt="Screen Shot 2022-09-23 at 12 06 49" src="https://user-images.githubusercontent.com/1680157/191992582-201038e5-9e9e-4ed3-870a-c34ce1823bf9.png">

Is it a thing that you're still addressing? Thanks!

gillesdemey commented:
That shouldn't have happened, I'll certainly try to address that – so far I've not observed this behavior when running it in the storybook but I'll check if I can repro it so I can come up with a fix or workaround.

mckn commented:
> That shouldn't have happened, I'll certainly try to address that – so far I've not observed this behavior when running it in the storybook but I'll check if I can repro it so I can come up with a fix or workaround.

Hmm, we have fixed this issue previously for the previous select implementation. Let me see if I can find the PR. It has to do with the width of the parent container.

mckn commented on packages/grafana-ui/src/components/Select/types.ts:
```diff
@@ -77,6 +77,8 @@ export interface SelectCommonProps<T> {
   tabSelectsValue?: boolean;
   value?: T | SelectValue<T> | null;
+  /** Will wrap the MenuList in a react-window FixedSizeVirtualList for improved performance, does not support options with "description" properties */
+  virtualized?: boolean;
```
Since this won't support options with descriptions I'm thinking that maybe it would be a good idea to expose this as its own Select component with a limited API instead of exporting it via the regular Select API (that currently is a mess).

gillesdemey replied:
Good idea, I'll see if I can refactor this to be a separate `VirtualizedSelect` component 👍

mckn replied:
Let me know if you need any assistance with that.

mckn reviewed (changes requested):
I found the [PR](https://github.com/grafana/grafana/pull/36398) that fixed this issue for the "regular" select. So what we need to do is to render the menu in a portal so it sits on top of the rest of the UI.

mckn commented:
> That shouldn't have happened, I'll certainly try to address that – so far I've not observed this behavior when running it in the storybook but I'll check if I can repro it so I can come up with a fix or workaround.

I think the best place to test this is in the panel edit UI.

gillesdemey commented:
> I found the https://github.com/grafana/grafana/pull/36398 that fixed this issue for the "regular" select. So what we need to do is to render the menu in a portal so it sits on top of the rest of the UI.

I'm not sure if this is the culprit, from what I can tell the Virtualized list is also rendered as a portal and attached to `document.body`. I think the issue comes from the virtual list using children that are `position: absolute;` so they're not making the parent container grow to accommodate their width 🤔

mckn commented:
> > I found the #36398 that fixed this issue for the "regular" select. So what we need to do is to render the menu in a portal so it sits on top of the rest of the UI.
> 
> I'm not sure if this is the culprit, from what I can tell the Virtualized list is also rendered as a portal and attached to `document.body`. I think the issue comes from the virtual list using children that are `position: absolute;` so they're not making the parent container grow to accommodate their width 🤔

Ah, @jackw pointed me in the following direction. 

https://github.com/grafana/grafana/pull/55629/files#diff-d14236403c8f563b9dffa1db4a3d8e5e130c4ebb05b7fc0436457aff7d9ea155R60

gillesdemey commented:
Thanks for the pointer @mckn – I've tried various values for the width including `auto` and `max-content` but none of them seem to have the right effect.

After looking a bit further it looks like the width of the drop-down is set by the `react-select` library, manually setting it to `width: auto;` makes the drop-down take up all of the available horizontal space.

I still believe this is because all of the child components in the virtual list are `position: absolute;` and so no space is created for those elements – making the container unable to know exactly who wide it needs to be to accommodate.

Perhaps the best we can do here is a "guesstimate" based on the longest value in the options (assuming items don't wrap)? 🤔

gillesdemey committed:
use a guesstimate for the menu width

gtk-grafana commented:
@gillesdemey Thanks for the work on this! I like your solution to determine a best guess for width, but wondering if we could create a width prop so in certain contexts where the automatic solution doesn't work, whoever is implementing the component can manually override the width?

obetomuniz commented:
> I've updated this PR with the guesstimate approach and it seems to work remarkably well – despite this obvious "hack" I invite you to try it out to see if it works.

Just tested the latest version of the PR, and it's working as expected. 

Without `virtualized` prop:
<img width="591" alt="Screen Shot 2022-09-27 at 15 47 07" src="https://user-images.githubusercontent.com/1680157/192610725-a798c670-1830-4431-9c9c-738d0c9bea16.png">


Using `virtualized` prop:
<img width="595" alt="Screen Shot 2022-09-27 at 15 46 09" src="https://user-images.githubusercontent.com/1680157/192610588-74313bcf-84ad-4b58-9c9d-0477781e3398.png">

gillesdemey committed:
use restricted interface for options in virtualized select component

virtualized supports pretty much every other "Select" prop but does not support every value from "SelectableValue"

gillesdemey commented:
@mckn I've exported a new `VirtualizedSelect` component that restricts the `options` that are supported, pretty much every other prop is compatible with the virtualized version so I'm not sure if there are more props I can omit from the `VirtualSelect` component.

This mean that you _might_ have to map an array of `SelectableValue<T>` to be compatible with the `VirtualizedSelect` component;

```js
// const options: Array<SelectableValue<T>> = [...]
const virtualizedOptions = options.map(({ label, value }) => ({
  value,
  label: label ?? value,
}))
```

mckn commented:
> @mckn I've exported a new `VirtualizedSelect` component that restricts the `options` that are supported, pretty much every other prop is compatible with the virtualized version so I'm not sure if there are more props I can omit from the `VirtualSelect` component.
> 
> This mean that you _might_ have to map an array of `SelectableValue<T>` to be compatible with the `VirtualizedSelect` component;
> 
> ```js
> // const options: Array<SelectableValue<T>> = [...]
> const virtualizedOptions = options.map(({ label, value }) => ({
>   value,
>   label: label ?? value,
> }))
> ```

Super! Great work on this!

It sounds like a less optimal side effect. I guess that is because of the type differences. What do you think about loosen the typings of the `VirtualSelect` so they are compatible with the `SelectableValue`?

mckn commented on packages/grafana-ui/src/components/Select/types.ts:
```diff
@@ -103,6 +105,14 @@ export interface SelectAsyncProps<T> {
 }
 
+/** The VirtualizedSelect component uses a slightly different SelectableValue, description and other props are not supported */
+export interface VirtualizedSelectProps<T> extends Omit<SelectCommonProps<T>, 'virtualized'> {
+  options?: Array<{
+    value: T;
+    label: string;
+  }>;
+}
```
```suggestion
export interface VirtualizedSelectProps<T> extends Omit<SelectCommonProps<T>, 'virtualized'> {
  options?: Array<{
    value?: T;
    label?: string;
  }>;
}
```

mckn replied:
Or maybe even something like this:

```suggestion
export interface VirtualizedSelectProps<T> extends Omit<SelectCommonProps<T>, 'virtualized'> {
  options?: Array<Pick<SelectableValue<T>, 'label' | 'value'>>;
}
```

gillesdemey replied:
Absolutely, great suggestion! 🙌

gillesdemey committed:
chore: loosen type definition

mckn reviewed (approved):
LGTM! Great work with this

gtk-grafana commented:
Hey, so I'm running into some trouble with this component crashing the browser when being passed a bunch (100k+) of options. I created a PR to reproduce the issue without requiring adding a bunch of data to prometheus: https://github.com/grafana/grafana/pull/57003/files.
To test: Checkout this branch, and then configure a prometheus datasource (or grab any existing), and then go to explore and click on the "Select metric" dropdown which will trigger the population of the 100k options.
<img width="634" alt="image" src="https://user-images.githubusercontent.com/109082771/195881672-1e731376-8af2-4835-9590-652de80d2540.png">

You'll see that it takes 15-30 seconds to render any list, and if you keep opening the component multiple times it will eventually crash the browser.

gtk-grafana commented:
Yeah I agree we need to rethink things to better allow for users that have potentially millions of metrics, this will likely necessitate backend changes so we can paginate and will take considerable engineering effort, so it's a good thing we're doing q4 planning right now! I'll create some issues for this.

mary-sitzenstatter commented:
Love this @gtk-grafana - I'm wondering if there is anyway for us to track the frequency of these types of errors with the metrics select box ... would love to have that data.

gtk-grafana commented:
We can't figure out how many times this error has been thrown since it uses all of the window's available memory and crashes (so no error reporting is possible), but we might be able see how many customers could be impacted by searching for all customers that have more then ~100k metrics (but they'd still need to be querying for a time interval that had 100k metrics reported to get this error, so might not be very accurate).

staton-hyse11 commented:
Reading through, I agree with @mary-sitzenstatter on this. It seems there is a misalignment between this control and the user's needs. Having that many options within a select is questionable and a poor experience trying to find a specific option. The search capability is helpful but I'm wondering if there is a better approach to this problem other than expanding our components to help with loading them.

The other thing I'm wondering is if we want to/need to set parameters around limits for this component to better control the usage of components like this. Is this how we want users to sort through thousands of metrics?

mary-sitzenstatter commented:
@staton-hyse11 - awesome comments, ty!

mary-sitzenstatter commented:
@gtk-grafana - is there any kind of log we could capture when that happens?
